### DISCUSSION TITLE
Using arbitrary values with `strictTokens: true` isn't possible

### DISCUSSION THREAD

**Discussion by bengry** [chakra-ui/panda#1001] [Ideas]
The `strictTokens` setting is great, and can catch typos like:
```tsx
<div className={css({ bg: 'red.5000' })} />
//                                ^ notice the extra 0
```

But with it turned on, using arbitrary values isn't possible. e.g.:
```tsx
<div className={css({
  maxWidth: '20ch'
}) />
```
[Fails to compile](https://play.panda-css.com/HMZLE3Ll7w).

I'm suggesting a syntax [similar to Tailwind](https://tailwindcss.com/docs/adding-custom-styles#using-arbitrary-values):
```tsx
<div className={css({
  bg: 'red.5000', // errors, not a valid token
  maxWidth: '[20ch]' // doesn't error, since it's inside `[]` - treated as an arbitrary value
}) />
```
Would generate:
```css
@layer utilities {
  .bg_red\.500 {
    background: var(--colors-red-500)
  }

  .max-w_20ch {
    max-width: 20ch;
  }
}
```

The exact syntax can be discussed, the main point is the possibility to use one-off styles when needed, without having to resort to turning off `strictTokens` entirely.

The TypeScript type to describe this would be:
```ts
type ArbitraryValue = `[${string | number}]`;
type TokenValue = /* Same thing you have today when strictTokens: true */;

type TokenOrArbitraryValue = TokenValue | ArbitraryValue;
```

This would be beneficial over just inline styles especially when integration with 3rd party libraries, which don't always allow passing in `style`, but almost always allow passing in `className` (or similar props).

FYI @astahmer

**Comment by bengry** [comment]
I think that the TypeScript part of this can be solved by changing [the definition of `PropertyValue`](https://github.com/chakra-ui/panda/blob/c53429d22f9e202c87714b833357494be6d3433a/packages/generator/src/artifacts/types/prop-types.ts#L46-L50):
```patch
  export type PropertyValue<T extends string> = 
+ (
	T extends keyof PropertyTypes
	    ? ConditionalValue<PropertyTypes[T]${strictText}>
	    : T extends keyof CssProperties
	    ? ConditionalValue<CssProperties[T]>
	    : ConditionalValue<string | number>
+ ) | ArbitraryValue;


+ type ArbitraryValue = \`[\${string | number\}]\`
```

I'm not sure if this is the most appropriate place to make this change, and not sure where's the code that extracts the value itself, but that would also need to be changed, of course, but I hope the above is a start at a solution to this.

**Comment by Pinpickle** [comment]
Another way to fix this is to create an identity function that satisfies the type system and Panda knows to analyse statically

```ts
// in styled-system/css

type ArbitraryValue = string & { __arbitraryValueBrand: "" }; // make this a branded type

// name could probably be improved
css.arbitrary = (value: string) => value as ArbitraryValue;
```

And then to use

```tsx
<div className={css({ width: css.arbitrary("5px") })} />
```

Like `css.raw`, Panda could special-case `css.arbitrary` and treat it as an identity function in static analysis.

**Comment by IHIutch** [comment]
Chiming in here to support this discussion (I started a similar one, https://github.com/chakra-ui/panda/discussions/909). As mentioned, turning on `strictTokens` is great for catching typos and mistakes. But it would also be nice to intentionally opt out for specific cases.

**Comment by neolectron** [comment]
It's related to [#1400 Strict tokens by token type in config](https://github.com/chakra-ui/panda/discussions/1400#discussioncomment-7447811), In my mind I see the best of both worlds like so :

Allow in the config :
```ts
{
...config,
tokenPolicies: {
  base: "strict",
  sizing: "relaxed",
  color: "semi-strict"
  }
}
```

Allow at usage :
```jsx
<div className={css({
  bg: 'red.500', // works in strict, it's in my config
  width: '123px', // works only because sizing is "relaxed" in the config.
  color: 'red', // works because color is "semi-strict" and red isn't part of my config but is a legit css value.
  __tokenPolicies: {
    relaxed: {
	  maxWidth: '20ch' // works because the arbitrary value is treated as "relaxed" because of the condition.
    }
  }
}) />
```

**Comment by hrsh7th** [comment]
Another use-case is here, I'm writing css with `content: '!'` but it doesn't allowed with `strictTokens: true`.

I strongly love the `strictTokens: true`. if it has strictness and flexibility, it would be a killer feature of panda :)

**Reply by hrsh7th** [reply]
The idea I thought is adding `{ allowedStyles: { content: ['!'] }` like configuration to panda.config.ts.

**Comment by astahmer** [comment]
Hey, we've been watching this thread and others on GH or Discord for a while and you're right, the DX with `config.strictTokens` can certainly be improved !
This [PR](https://github.com/chakra-ui/panda/pull/1756) will add an escape-hatch for arbitrary values using the `[xxx]` syntax
